extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -0,0 +1,334 @@
+   * all bindings.
+   * @param {!Element} body
+   * @return {!Array<BindingDef>}
+   * @private
+   */
+  scanForBindings_(body) {
+    // TODO(choumx): Chunk if taking too long in a single frame.
+    const bindings = [];
+    const elements = body.getElementsByTagName('*');
+    for (let i = 0; i < elements.length; i++) {
+      const el = elements[i];
+      const attributes = el.attributes;
+      for (let j = 0; j < attributes.length; j++) {
+        const binding = this.bindingForAttribute_(attributes[j], el);
+        if (binding) {
+          bindings.push(binding);
+        }
+      }
+    }
+    return bindings;
+  }
+
+  /**
+   * Returns a struct representing the binding corresponding to the
+   * attribute param, if applicable.
+   * @param {!Attr} attribute
+   * @param {!Element} element
+   * @return {?BindingDef}
+   * @private
+   */
+  bindingForAttribute_(attribute, element) {
+    // TODO(choumx): Only allow binding to attributes allowed by validator.
+
+    const name = attribute.name;
+    if (name.length > 2) {
+      if (name[0] === '[' && name[name.length - 1] === ']') {
+        return {
+          property: name.substr(1, name.length - 2),
+          expression: attribute.value,
+          element,
+        };
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Schedules a vsync task to reevaluate all binding expressions.
+   * @param {boolean=} opt_verifyOnly
+   * @private
+   */
+  digest_(opt_verifyOnly) {
+    // TODO(choumx): Chunk if takes too long for a single frame.
+
+    /** {!VsyncTaskSpecDef} */
+    const task = {measure: this.boundMeasure_, mutate: this.boundMutate_};
+    this.vsync_.run(task, {
+      results: [],
+      verifyOnly: opt_verifyOnly,
+    });
+  }
+
+  /**
+   * Reevaluates all bindings and stores results in `state`.
+   * @param {BindVsyncStateDef} state
+   * @private
+   */
+  measure_(state) {
+    for (let i = 0; i < this.bindings_.length; i++) {
+      const binding = this.bindings_[i];
+      state.results[i] = evaluateBindExpr(binding.expression, this.scope_);
+    }
+  }
+
+  /**
+   * Either applies or verifies the binding evaluation results in `state`.
+   * @param {BindVsyncStateDef} state
+   * @private
+   */
+  mutate_(state) {
+    for (let i = 0; i < this.bindings_.length; i++) {
+      const binding = this.bindings_[i];
+      const result = state.results[i];
+      if (state.verifyOnly) {
+        this.verifyBinding_(binding, result);
+      } else {
+        this.applyBinding_(binding, result);
+      }
+    }
+  }
+
+  /**
+   * Applies `newValue` to the element bound in `binding`.
+   * @param {!BindingDef} binding
+   * @param {BindExpressionResultDef} newValue
+   * @private
+   */
+  applyBinding_(binding, newValue) {
+    const property = binding.property;
+    const element = binding.element;
+
+    // TODO(choumx): Support objects for attributes.
+
+    if (property === 'text') {
+      element.textContent = newValue;
```

Comment:
Think we ought to find and mutate the child text node instead? That may be more user friendly but I'm a bit concerned about performance implications.

Mutating `textContent` also doesn't trigger reflow which users will probably want. Any suggestions on how to handle that?